Merge 4.2 dev to master

ibgreen opened the pull request:
I've done a lot of work on polishing the 4.2-dev and I believe this is a good time to merge.

shaojingli commented:
There are conflicts between these two branches and also can you specify which tests have this branch gone through?

ibgreen commented:
> There are conflicts between these two branches

Fixed. Another PR landed on master after I rebased 4.2-dev. At this point most PRs that land on master could introduce some conflict, but the big rebase has been done.

> and also can you specify which tests have this branch gone through?

* layer browser testing
* Also carefully went through our test suite and replaced the old layer test utils with code that uses the full layer manager logic instead of just snippets of code. Makes the test suite exercise the same code as during runtime, and make it less fragile when refactoring life cycle functionality.

ibgreen commented:
Tested this in internal app as well, works well. Given that component PRs are all approved I will go ahead and merging.
